src/Common/SecurityToken.cpp

```diff
@@ -361,7 +576,115 @@ namespace VeraCrypt
 		return objects;
 	}
 
-	void SecurityToken::GetObjectAttribute(CK_SLOT_ID slotId, CK_OBJECT_HANDLE tokenObject, CK_ATTRIBUTE_TYPE attributeType, vector <uint8>& attributeValue)
+
+	CK_RV SecurityTokenImpl::PKCS11Encrypt(
+		CK_SESSION_HANDLE hSession,
+		vector<uint8> plaintext,
+		vector<uint8> &ciphertext
+		)
+	{
+		CK_RV rv;
+		if (!plaintext.size())
+			return CKR_ARGUMENTS_BAD;
+
+		CK_ULONG ulInDataLen = plaintext.size();
+		CK_BYTE* pInData = plaintext.data();
+		CK_ULONG ulOutDataLen = ciphertext.size();
+		CK_BYTE* pOutData = ciphertext.data();
+
+		rv = Pkcs11Functions->C_Encrypt(hSession, pInData, ulInDataLen, pOutData,
+			&ulOutDataLen);
+
+		if (CKR_OK == rv) {
+			ciphertext = vector<uint8>(pOutData, pOutData + ulOutDataLen);
+		} else {
+			throw Pkcs11Exception(rv);
+		}
+		return rv;
+	}
+
+	CK_RV SecurityTokenImpl::PKCS11Decrypt(
+		CK_SESSION_HANDLE hSession,
+		vector<uint8> inEncryptedData,
+		vector<uint8> &outData
+		)
+	{
+		CK_RV rv;
+		if (!inEncryptedData.size())
+			return CKR_ARGUMENTS_BAD;
+
+		CK_ULONG ulInDataLen = inEncryptedData.size();
+		CK_BYTE* pInData = inEncryptedData.data();
+		CK_ULONG ulOutDataLen = outData.size();
+		CK_BYTE* pOutData = outData.data();
+
+		rv = Pkcs11Functions->C_Decrypt(hSession, pInData, ulInDataLen, pOutData,
+			&ulOutDataLen);
+
+		if (CKR_OK == rv) {
+			outData = vector<uint8>(pOutData, pOutData + ulOutDataLen);
+		} else {
+			throw Pkcs11Exception(rv);
+		}
+		return rv;
+	}
+
+	void SecurityTokenImpl::GetEncryptedData(SecurityTokenKey key, vector<uint8> plaintext, vector<uint8> &ciphertext) {
+		GetEncryptedData(key.SlotId, key.Handle, plaintext, ciphertext);
+	}
+
+	void SecurityTokenImpl::GetEncryptedData (CK_SLOT_ID slotId, CK_OBJECT_HANDLE tokenObject, vector <uint8> plaintext, vector <uint8> &ciphertext)
+	{
+		LoginUserIfRequired (slotId);
+
+		if (Sessions.find (slotId) == Sessions.end())
+			throw ParameterIncorrect (SRC_POS);
+
+		CK_MECHANISM m = { CKM_RSA_PKCS, 0, 0 };
+		CK_RV status = Pkcs11Functions->C_EncryptInit (Sessions[slotId].Handle, &m, tokenObject);
```

Comment:
We should not use CKM_RSA_PKCS since it is insecure. We must use CKM_RSA_PKCS_OAEP with SHA256 MGF.